File: app/src/main/java/com/motorola/cdeives/motofretado/TrackBusPresenter.java
package com.motorola.cdeives.motofretado;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Message;
import android.os.Messenger;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.UiThread;
import android.text.TextUtils;
import android.util.Log;

import com.motorola.cdeives.motofretado.http.Bus;
import com.motorola.cdeives.motofretado.http.ModelListener;

import java.lang.ref.WeakReference;
import java.util.Arrays;
import java.util.List;

class TrackBusPresenter implements TrackBusMvp.Presenter {
    private static final String TAG = TrackBusPresenter.class.getSimpleName();
    private static final String MOST_RECENT_TRACK_BUS_ID_PREF = "most_recent_track_bus_id";

    private final @NonNull Context mContext;
    private final @NonNull TrackBusMvp.Model mModel;
    private @Nullable TrackBusMvp.View mView;
    private final @NonNull Intent mActivityDetectionServiceIntent;
    private final @NonNull Intent mUpdateLocationServiceIntent;
    private boolean mIsUpdateLocationServiceRunning;
    private boolean mIsActivityDetectionServiceRunning;
    private @Nullable String mSelectedBusId;
    private @Nullable List<Bus> mAvailableBuses;

    @UiThread
    TrackBusPresenter(@NonNull Context context) {
        mContext = context;

        mModel = new TrackBusModel(mContext);
        Messenger messenger = new Messenger(new MyHandler(this));

        mActivityDetectionServiceIntent = new Intent(mContext, ActivityDetectionService.class);
        mActivityDetectionServiceIntent.putExtra(
                ActivityDetectionService.EXTRA_MESSENGER, messenger);

        mUpdateLocationServiceIntent = new Intent(mContext, UpdateLocationService.class);
        mUpdateLocationServiceIntent.putExtra(UpdateLocationService.EXTRA_MESSENGER, messenger);

        mModel.readAllBuses(new ReadAllBusesListener(null,true));
    }

    @Override
    @UiThread
    public void onAttach(@NonNull TrackBusMvp.View view) {
        mView = view;

        if (mAvailableBuses != null) {
            mView.setAvailableBuses(mAvailableBuses, mSelectedBusId);
            if (!mAvailableBuses.isEmpty()) {
                mView.enableBusId();
            }
        }

        if (mIsUpdateLocationServiceRunning || mIsActivityDetectionServiceRunning) {
            mView.disableBusId();
        }

        if (!mIsActivityDetectionServiceRunning) {
            mView.uncheckSwitchDetectAutomatically();
        }
    }

    @Override
    @UiThread
    public void onDetach() {
        if (mView != null) {
            mSelectedBusId = mView.getBusId();
        } else {
            Log.w(TAG, "view is null; cannot save current bus ID");
        }

        mView = null;
    }

    @Override
    @UiThread
    public void startLocationUpdate() {
        if (mView != null) {
            String busId = mView.getBusId();

            Log.d(TAG, "starting service " + mUpdateLocationServiceIntent.getComponent());
            mUpdateLocationServiceIntent.putExtra(UpdateLocationService.EXTRA_BUS_ID, busId);
            mContext.startService(mUpdateLocationServiceIntent);

            Log.d(TAG, "writing preference: " + MOST_RECENT_TRACK_BUS_ID_PREF + " => " + busId);
            SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(mContext);
            SharedPreferences.Editor prefsEditor = sharedPrefs.edit();
            prefsEditor.putString(MOST_RECENT_TRACK_BUS_ID_PREF, busId);
            prefsEditor.apply();
        } else {
            Log.w(TAG, "view is null; cannot read bus ID needed to start the location service");
        }
    }

    @Override
    @UiThread
    public void stopLocationUpdate() {
        if (mIsUpdateLocationServiceRunning) {
            Log.d(TAG, "stopping service " + mUpdateLocationServiceIntent.getComponent());
            mContext.stopService(mUpdateLocationServiceIntent);
        } else {
            Log.d(TAG, "service is already stopped; there's no need to stop it");
        }
    }

    @Override
    @UiThread
    public void startActivityDetection() {
        Log.d(TAG, "starting service " + mActivityDetectionServiceIntent.getComponent());

        if (mView != null) {
            mView.disableBusId();
        } else {
            Log.w(TAG, "view is null; cannot disable the bus ID field");
        }

        mContext.startService(mActivityDetectionServiceIntent);
    }

    @Override
    @UiThread
    public void stopActivityDetection() {
        Log.d(TAG, "stopping service " + mActivityDetectionServiceIntent.getComponent());
        mContext.stopService(mActivityDetectionServiceIntent);
    }

    @Override
    @UiThread
    public void createBus(@NonNull Bus bus) {
        Log.d(TAG, "creating bus: " + bus);

        mModel.createBus(bus, new PostBusListener());
    }

    static class MyHandler extends Handler {
        static final int MSG_DISPLAY_TOAST = 0;
        static final int MSG_GMS_CONNECTION_FAILED = 1;
        static final int MSG_UPDATE_LOCATION_SERVICE_CONNECTED = 2;
        static final int MSG_UPDATE_LOCATION_SERVICE_DISCONNECTED = 3;
        static final int MSG_ACTIVITY_DETECTION_SERVICE_CONNECTED = 4;
        static final int MSG_ACTIVITY_DETECTION_SERVICE_DISCONNECTED = 5;
        static final int MSG_USER_IS_ON_FOOT = 6;
        static final int MSG_USER_IS_IN_VEHICLE = 7;

        private final @NonNull WeakReference<TrackBusPresenter> mPresenterRef;

        MyHandler(TrackBusPresenter presenter) {
            mPresenterRef = new WeakReference<>(presenter);
        }

        @Override
        @UiThread
        public void handleMessage(Message msg) {
            Log.v(TAG, "> handleMessage(msg=" + msg + ")");

            TrackBusPresenter presenter = mPresenterRef.get();
            if (presenter != null) {
                switch (msg.what) {
                    case MSG_DISPLAY_TOAST:
                        if (presenter.mView != null) {
                            presenter.mView.displayMessage(presenter.mContext.getString(msg.arg1));
                        } else {
                            Log.w(TAG, "view is null; cannot display message");
                        }
                        break;
                    case MSG_GMS_CONNECTION_FAILED:
                        if (presenter.mView != null) {
                            presenter.mView.displayMessage(
                                    presenter.mContext.getString(R.string.gms_connection_failed));
                        } else {
                            Log.d(TAG, "view is null; cannot display message");
                        }
                        break;
                    case MSG_UPDATE_LOCATION_SERVICE_DISCONNECTED:
                        presenter.mIsUpdateLocationServiceRunning = false;
                        presenter.mIsActivityDetectionServiceRunning = false;
                        if (presenter.mView != null) {
                            presenter.mView.enableBusId();
                            presenter.mView.uncheckSwitchDetectAutomatically();
                        } else {
                            Log.d(TAG, "view is null; cannot update UI");
                        }
                        presenter.mModel.cancelAllRequests();
                        break;
                    case MSG_UPDATE_LOCATION_SERVICE_CONNECTED:
                        presenter.mIsUpdateLocationServiceRunning = true;
                        if (presenter.mView != null) {
                            presenter.mView.disableBusId();
                        } else {
                            Log.d(TAG, "view is null; cannot update UI");
                        }
                        break;
                    case MSG_ACTIVITY_DETECTION_SERVICE_CONNECTED:
                        presenter.mIsActivityDetectionServiceRunning = true;
                        break;
                    case MSG_ACTIVITY_DETECTION_SERVICE_DISCONNECTED:
                        presenter.mIsActivityDetectionServiceRunning = false;
                        if (presenter.mView != null) {
                            if (!presenter.mIsUpdateLocationServiceRunning) {
                                presenter.mView.enableBusId();
                            }
                        } else {
                            Log.d(TAG, "view is null; cannot update UI");
                        }
                        break;
                    case MSG_USER_IS_ON_FOOT:
                        if (presenter.mIsUpdateLocationServiceRunning) {
                            presenter.stopLocationUpdate();
                            presenter.stopActivityDetection();
                        }
                        break;
                    case MSG_USER_IS_IN_VEHICLE:
                        if (!presenter.mIsUpdateLocationServiceRunning) {
                            presenter.startLocationUpdate();
                        }
                        break;
                    default:
                        Log.wtf(TAG, "unexpected message code: " + msg.what);
                }
            } else {
                Log.w(TAG, "presenter reference doesn't exist anymore; ignoring message");
            }

            Log.v(TAG, "< handleMessage(msg=" + msg + ")");
        }
    }

    @UiThread
    private class ReadAllBusesListener implements ModelListener<Bus[]> {
        private final @Nullable String mSelectedBusId;
        private final boolean mSelectDefaultFromPref;

        private ReadAllBusesListener(@Nullable String selectedBusId, boolean selectedBusFromPref) {
            mSelectedBusId = selectedBusId;
            mSelectDefaultFromPref = selectedBusFromPref;
        }

        @Override
        public void onSuccess(Bus[] buses) {
            List<Bus> busesList = Arrays.asList(buses);

            if (mView != null) {
                String selectedBusId;

                if (!busesList.isEmpty()) {
                    if (mSelectDefaultFromPref) {
                        selectedBusId = PreferenceManager.getDefaultSharedPreferences(mContext)
                                .getString(MOST_RECENT_TRACK_BUS_ID_PREF, null);
                        Log.d(TAG, "preference read: "
                                + MOST_RECENT_TRACK_BUS_ID_PREF + " => " + selectedBusId);
                    } else if (!TextUtils.isEmpty(mSelectedBusId)) {
                        selectedBusId = mSelectedBusId;
                    } else {
                        selectedBusId = null;
                    }

                    mView.enableBusId();
                } else {
                    selectedBusId = null;
                }

                mView.setAvailableBuses(busesList, selectedBusId);
                mAvailableBuses = busesList;
            } else {
                Log.w(TAG, "view is null; cannot update the available bus numbers");
            }
        }

        @Override
        public void onError(Exception ex) {
            Log.e(TAG, "could not read buses", ex);
            if (mView != null) {
                mView.displayMessage(mContext.getString(R.string.read_buses_failed));
            } else {
                Log.w(TAG, "view is null; cannot display error message");
            }
        }
    }

    @UiThread
    private class PostBusListener implements ModelListener<Bus> {
        @Override
        public void onSuccess(Bus bus) {
            Log.d(TAG, "bus created successfully: " + bus);
            mModel.readAllBuses(new ReadAllBusesListener(bus.id, false));
        }

        @Override
        public void onError(Exception ex) {
            Log.e(TAG, "could not create bus", ex);
            if (mView != null) {
                mView.displayMessage(mContext.getString(R.string.create_bus_failed));
            } else {
                Log.w(TAG, "view is null; cannot display error message");
            }
        }
    }
}

File: app/src/main/java/com/gmail/cristiandeives/motofretado/TrackBusPresenter.kt
package com.gmail.cristiandeives.motofretado

import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.Message
import android.os.Messenger
import android.preference.PreferenceManager
import android.support.annotation.UiThread
import android.util.Log
import com.gmail.cristiandeives.motofretado.http.Bus
import com.gmail.cristiandeives.motofretado.http.ModelListener
import java.lang.ref.WeakReference

@UiThread
internal class TrackBusPresenter(private val mContext: Context) : TrackBusMvp.Presenter {
    companion object {
        private val TAG = TrackBusPresenter::class.java.simpleName
        private const val MOST_RECENT_TRACK_BUS_ID_PREF = "most_recent_track_bus_id"
    }

    private val mModel: TrackBusMvp.Model = TrackBusModel(mContext)
    private var mView: TrackBusMvp.View? = null
    private val mActivityDetectionServiceIntent: Intent
    private val mUpdateLocationServiceIntent: Intent
    private var mIsUpdateLocationServiceRunning: Boolean = false
    private var mIsActivityDetectionServiceRunning: Boolean = false
    private var mSelectedBusId: String? = null
    private var mAvailableBuses: List<Bus>? = null
    private var mBusErrorMessage: String? = null

    init {
        val messenger = Messenger(MyHandler(this))

        mActivityDetectionServiceIntent = Intent(mContext, ActivityDetectionService::class.java).apply {
            putExtra(ActivityDetectionService.EXTRA_MESSENGER, messenger)
        }

        mUpdateLocationServiceIntent = Intent(mContext, UpdateLocationService::class.java).apply {
            putExtra(UpdateLocationService.EXTRA_MESSENGER, messenger)
        }

        mModel.readAllBuses(ReadAllBusesListener(null, true))
    }

    override fun onAttach(view: TrackBusMvp.View) {
        mView = view

        val errorMessage = mBusErrorMessage
        if (errorMessage != null && errorMessage.isNotEmpty()) {
            view.setBusError(errorMessage)
        } else {
            mAvailableBuses?.let { buses ->
                view.apply {
                    setAvailableBuses(buses, mSelectedBusId)
                    enableBusId()
                }
            }
        }

        if (mIsUpdateLocationServiceRunning || mIsActivityDetectionServiceRunning) {
            view.disableBusId()
        }

        if (!mIsActivityDetectionServiceRunning) {
            view.uncheckSwitchDetectAutomatically()
        }
    }

    override fun onDetach() {
        mSelectedBusId = mView?.getBusId()
        mView = null
    }

    override fun startLocationUpdate() {
        mView?.let { view ->
            val busId = view.getBusId()

            Log.d(TAG, "starting service ${mUpdateLocationServiceIntent.component}")
            mUpdateLocationServiceIntent.putExtra(UpdateLocationService.EXTRA_BUS_ID, busId)
            mContext.startService(mUpdateLocationServiceIntent)

            Log.d(TAG, "writing preference: $MOST_RECENT_TRACK_BUS_ID_PREF => $busId")
            mContext.editSharedPreferences {
                putString(MOST_RECENT_TRACK_BUS_ID_PREF, busId)
            }
        }
    }

    override fun stopLocationUpdate() {
        if (mIsUpdateLocationServiceRunning) {
            Log.d(TAG, "stopping service ${mUpdateLocationServiceIntent.component}")
            mContext.stopService(mUpdateLocationServiceIntent)
        }
    }

    override fun startActivityDetection() {
        Log.d(TAG, "starting service ${mActivityDetectionServiceIntent.component}")

        mView?.disableBusId()
        mContext.startService(mActivityDetectionServiceIntent)
    }

    override fun stopActivityDetection() {
        Log.d(TAG, "stopping service ${mActivityDetectionServiceIntent.component}")
        mContext.stopService(mActivityDetectionServiceIntent)
    }

    override fun createBus(bus: Bus) {
        Log.d(TAG, "creating bus: $bus")
        mModel.createBus(bus, PostBusListener())
    }

    internal class MyHandler(presenter: TrackBusPresenter) : Handler() {
        companion object {
            const val MSG_DISPLAY_TOAST = 0
            const val MSG_GMS_CONNECTION_FAILED = 1
            const val MSG_UPDATE_LOCATION_SERVICE_CONNECTED = 2
            const val MSG_UPDATE_LOCATION_SERVICE_DISCONNECTED = 3
            const val MSG_ACTIVITY_DETECTION_SERVICE_CONNECTED = 4
            const val MSG_ACTIVITY_DETECTION_SERVICE_DISCONNECTED = 5
            const val MSG_USER_IS_ON_FOOT = 6
            const val MSG_USER_IS_IN_VEHICLE = 7
        }

        private val mPresenterRef: WeakReference<TrackBusPresenter> = WeakReference(presenter)

        override fun handleMessage(msg: Message) {
            Log.v(TAG, "> handleMessage(msg=$msg)")

            mPresenterRef.get()?.let { presenter ->
                when (msg.what) {
                    MSG_DISPLAY_TOAST -> presenter.mView?.displayMessage(presenter.mContext.getString(msg.arg1))
                    MSG_GMS_CONNECTION_FAILED -> presenter.mView?.displayMessage(presenter.mContext.getString(R.string.gms_connection_failed))
                    MSG_UPDATE_LOCATION_SERVICE_DISCONNECTED -> {
                        presenter.apply {
                            mIsUpdateLocationServiceRunning = false
                            mIsActivityDetectionServiceRunning = false
                            mView?.let { view ->
                                view.enableBusId()
                                view.uncheckSwitchDetectAutomatically()
                            }
                            mModel.cancelAllRequests()
                        }
                    }
                    MSG_UPDATE_LOCATION_SERVICE_CONNECTED -> {
                        presenter.apply {
                            mIsUpdateLocationServiceRunning = true
                            mView?.disableBusId()
                        }
                    }
                    MSG_ACTIVITY_DETECTION_SERVICE_CONNECTED -> presenter.mIsActivityDetectionServiceRunning = true
                    MSG_ACTIVITY_DETECTION_SERVICE_DISCONNECTED -> presenter.apply {
                        mIsActivityDetectionServiceRunning = false
                        if (!mIsUpdateLocationServiceRunning) {
                            mView?.enableBusId()
                        }
                    }
                    MSG_USER_IS_ON_FOOT -> presenter.apply {
                        if (mIsUpdateLocationServiceRunning) {
                            stopLocationUpdate()
                            stopActivityDetection()
                        }
                    }
                    MSG_USER_IS_IN_VEHICLE -> presenter.apply {
                        if (!mIsUpdateLocationServiceRunning) {
                            startLocationUpdate()
                        }
                    }
                    else -> Log.wtf(TAG, "unexpected message code: ${msg.what}")
                }
            }

            Log.v(TAG, "< handleMessage(msg=$msg)")
        }
    }

    private inner class ReadAllBusesListener(private val mSelectedBusId: String?, private val mSelectDefaultFromPref: Boolean) : ModelListener<List<Bus>> {
        override fun onSuccess(data: List<Bus>) {
            mView?.let { view ->
                if (!data.isEmpty()) {
                    val selectedBusId = if (mSelectDefaultFromPref) {
                        val busIdPref = PreferenceManager.getDefaultSharedPreferences(mContext)
                                .getString(MOST_RECENT_TRACK_BUS_ID_PREF, null)
                        Log.d(TAG, "preference read: $MOST_RECENT_TRACK_BUS_ID_PREF => $busIdPref")
                        busIdPref
                    } else if (!mSelectedBusId.isNullOrEmpty()) {
                        mSelectedBusId
                    } else {
                        null
                    }

                    view.enableBusId()
                    view.setAvailableBuses(data, selectedBusId)
                    mBusErrorMessage = null
                    mAvailableBuses = data
                } else {
                    val errorMessage = mContext.getString(R.string.no_buses_available)
                    view.setBusError(errorMessage)
                    mBusErrorMessage = errorMessage
                }
            }
        }

        override fun onError(ex: Exception) {
            Log.e(TAG, "could not read buses", ex)
            val errorMessage = mContext.getString(R.string.read_buses_failed)
            mView?.setBusError(errorMessage)
            mBusErrorMessage = errorMessage
        }
    }

    private inner class PostBusListener : ModelListener<Bus> {
        override fun onSuccess(data: Bus) {
            Log.d(TAG, "bus created successfully: $data")
            mModel.readAllBuses(ReadAllBusesListener(data.id, false))
        }

        override fun onError(ex: Exception) {
            Log.e(TAG, "could not create bus", ex)
            mView?.displayMessage(mContext.getString(R.string.create_bus_failed))
        }
    }
}